File: backend/app/models/student_term_record.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

AcademicStanding = Literal[
    "good",
    "warning",
    "probation",
    "suspension",
]


class StudentTermRecordBase(BaseModel):
    person_id: UUID
    term_id: UUID
    program_id: UUID | None = None
    credits_attempted: Decimal = Decimal("0")
    credits_earned: Decimal = Decimal("0")
    term_gpa: Decimal | None = Field(
        default=None,
        ge=0,
        le=4,
    )
    cumulative_gpa: Decimal | None = Field(
        default=None,
        ge=0,
        le=4,
    )
    academic_standing: AcademicStanding | None = None
    advisor_meetings: int = 0
    attributes: dict[str, Any] = Field(default_factory=dict)


class StudentTermRecordCreate(StudentTermRecordBase):
    pass


class StudentTermRecordUpdate(BaseModel):
    program_id: UUID | None = None
    credits_attempted: Decimal | None = None
    credits_earned: Decimal | None = None
    term_gpa: Decimal | None = Field(
        default=None,
        ge=0,
        le=4,
    )
    cumulative_gpa: Decimal | None = Field(
        default=None,
        ge=0,
        le=4,
    )
    academic_standing: AcademicStanding | None = None
    advisor_meetings: int | None = None
    attributes: dict[str, Any] | None = None


class StudentTermRecordResponse(StudentTermRecordBase):
    id: UUID
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/models/studentTermRecord.js
// src/models/studentTermRecord.js

class StudentTermRecord {
  constructor(
    id,
    person_id,
    term_id,
    program_id,
    credits_attempted,
    credits_earned,
    term_gpa,
    cumulative_gpa,
    academic_standing,
    advisor_meetings,
    attributes,
    created_at
  ) {
    this.id = id;
    this.person_id = person_id;
    this.term_id = term_id;
    this.program_id = program_id;
    this.credits_attempted = credits_attempted;
    this.credits_earned = credits_earned;
    this.term_gpa = term_gpa;
    this.cumulative_gpa = cumulative_gpa;
    this.academic_standing = academic_standing;
    this.advisor_meetings = advisor_meetings;
    this.attributes = attributes;
    this.created_at = created_at;
  }

  // Helper getters for camelCase compatibility
  get personId() {
    return this.person_id;
  }

  get termId() {
    return this.term_id;
  }

  get programId() {
    return this.program_id;
  }

  get creditsAttempted() {
    return this.credits_attempted;
  }

  get creditsEarned() {
    return this.credits_earned;
  }

  get termGpa() {
    return this.term_gpa;
  }

  get cumulativeGpa() {
    return this.cumulative_gpa;
  }

  get academicStanding() {
    return this.academic_standing;
  }

  get advisorMeetings() {
    return this.advisor_meetings;
  }

  get createdAt() {
    return this.created_at;
  }

  static fromApiResponse(data) {
    if (!data) return null;
    return new StudentTermRecord(
      data.id,
      data.person_id ?? data.personId,
      data.term_id ?? data.termId,
      data.program_id ?? data.programId,
      data.credits_attempted ?? data.creditsAttempted,
      data.credits_earned ?? data.creditsEarned,
      data.term_gpa ?? data.termGpa,
      data.cumulative_gpa ?? data.cumulativeGpa,
      data.academic_standing ?? data.academicStanding,
      data.advisor_meetings ?? data.advisorMeetings,
      data.attributes,
      data.created_at ?? data.createdAt
    );
  }

  toJSON() {
    return {
      id: this.id,
      person_id: this.person_id,
      term_id: this.term_id,
      program_id: this.program_id,
      credits_attempted: this.credits_attempted,
      credits_earned: this.credits_earned,
      term_gpa: this.term_gpa,
      cumulative_gpa: this.cumulative_gpa,
      academic_standing: this.academic_standing,
      advisor_meetings: this.advisor_meetings,
      attributes: this.attributes,
      created_at: this.created_at,
    };
  }

  createPayload() {
    return {
      person_id: this.person_id,
      term_id: this.term_id,
      program_id: this.program_id,
      credits_attempted: this.credits_attempted,
      credits_earned: this.credits_earned,
      term_gpa: this.term_gpa,
      cumulative_gpa: this.cumulative_gpa,
      academic_standing: this.academic_standing,
      advisor_meetings: this.advisor_meetings,
      attributes: this.attributes,
    };
  }
}

export default StudentTermRecord;
